pow/src/main/java/tech/pegasys/artemis/pow/Eth1DataManager.java

```diff
@@ -0,0 +1,372 @@
+
+Definitions:
+  t: current time
+  Request Buffer: the time it will take for potential eth1 data request to complete
+
+Constants:
+  Eth1 Follow Distance Time = Seconds Per Eth1 Block * Eth1 Follow Distance
+  Slots Per Eth1 Voting Period Time = Seconds Per Slot * Slots Per Eth1 Voting Period
+
+cache range =
+((t - (Slots Per Eth1 Voting Period Time) - (Eth1 Follow Distance Time * 2),
+(t - (Eth1 Follow Distance Time) + Request Buffer))
+
+At startup: (implemented in this class)
+  - Find the cache range
+  - Search Eth1 blocks to find blocks in the cache range (pseudo-code defined below)
+
+On every Slot Event: (implemented in this class)
+  - Get the latest block number you have
+  - Calculate upper bound, i.e.  (t - (Eth1 Follow Distance Time) + Request Buffer))
+  - Request blocks from i = 0 to  i = to infinity, until latest block number + i’s timestamp
+  is greater than the upper bound
+
+On every VotingPeriodStart change: (implemented in Eth1DataCache)
+  - Prune anything that is before than:
+  ((t - (Slots Per Eth1 Voting Period Time - One Slot Time) - (Eth1 Follow Distance Time * 2)
+
+Search Eth1 Blocks to find blocks in the cache range:
+  1) Get the latest block’s time stamp and block number
+  2) rcr_average = ((rcr_lower_bound + rcr_upper_bound) / 2)
+  3) time_diff = latest_block_timestamp - rcr_average
+  4) seconds_per_eth1_block = SECONDS_PER_ETH1_BLOCK
+  5) block_number_diff = time_diff / seconds_per_eth1_block
+  6) block_number = latest_block_number - block_number_diff
+  7) block_timestamp = getEthBlock(block_number).timestamp
+  8) if isTimestampInRCR(block_timestamp):
+      - go in both directions until you’re not in the range
+      - post each block to event bus
+     else:
+      - actual_time_diff = latest_block_timestamp - block_timestamp
+      - seconds_per_eth1_block = block_number_diff / actual_time_diff
+      - go back to step 5
+
+ */
+
+public class Eth1DataManager {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Web3j web3j;
+  private final DepositContractListener depositContractListener;
+  private final EventBus eventBus;
+
+  private AtomicReference<EthBlock.Block> latestBlockReference = new AtomicReference<>();
+
+  private enum StartupLogicStates {
+    SUCCESSFULLY_COMPLETED,
+    UNSUCCESSFULLY_COMPLETED,
+    UNABLE_TO_EXPLORE_BLOCKS,
+    DONE_EXPLORING,
+  }
+
+  public Eth1DataManager(
+      Web3j web3j, EventBus eventBus, DepositContractListener depositContractListener) {
+    this.web3j = web3j;
+    this.depositContractListener = depositContractListener;
+    this.eventBus = eventBus;
+    eventBus.register(this);
+
+    runCacheStartupLogic()
+        .finish(
+            (result) -> {
+              if (!result.equals(StartupLogicStates.SUCCESSFULLY_COMPLETED)) {
+                throw new RuntimeException("Eth1DataManager unable to fill cache at startup");
+              }
+            });
```

Comment:
We probably should have `runCacheStartupLogic` just return a `SafeFuture<Void>` and if it fails the `SafeFuture` completes exceptionally.  Then here we can catch that exception (with either exceptionally or finish(onSuccess, onFailure)) and retry after a short delay.  If we never retry then artemis will lose connection with the Eth1 chain and need to be restarted before it recovers.

Reply:
Done. Actually had to make runCacheStartupLogic implement .finish() in itself due to needing to schedule the retry attempts multiple times in the future, so runCacheStartupLogic returns void in the end. Hopefully, that was the right thing to do.